src/polymer/expression-scanner.ts

```diff
@@ -0,0 +1,163 @@
+/**
+ * @license
+ * Copyright (c) 2016 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import * as dom5 from 'dom5';
+import * as espree from 'espree';
+import * as estree from 'estree';
+import * as parse5 from 'parse5';
+
+import {HtmlVisitor, ParsedHtmlDocument} from '../html/html-document';
+import {HtmlScanner} from '../html/html-scanner';
+import {baseParseOptions} from '../javascript/javascript-parser';
+import {ScannedFeature, SourceRange} from '../model/model';
+import {Warning} from '../warning/warning';
+
+const p = dom5.predicates;
+const isTemplate = p.hasTagName('template');
+const matchInnerDataBindingTemplate = p.AND(
+    isTemplate,
+    p.OR(
+        p.hasAttrValue('is', 'dom-bind'),
+        p.hasAttrValue('is', 'dom-if'),
+        p.hasAttrValue('is', 'dom-repeat'),
+        p.parentMatches(p.OR(
+            p.hasTagName('dom-bind'),
+            p.hasTagName('dom-if'),
+            p.hasTagName('dom-repeat')))));
+export const isDomBindTemplate =
+    p.AND(p.hasTagName('template'), p.hasAttrValue('is', 'dom-bind'));
+export const isDomModuleTemplate = p.AND(
+    p.parentMatches(p.hasTagName('dom-module')), p.hasTagName('template'));
+
+export interface Template extends parse5.ASTNode { content: parse5.ASTNode; }
+
+export function getAllDataBindingTemplates(node: parse5.ASTNode) {
+  const toplevelTemplates = dom5.queryAll(
+      node,
+      dom5.predicates.OR(isDomBindTemplate, isDomModuleTemplate)) as Template[];
+
+  const results = new Set<Template>();
+  for (const template of toplevelTemplates) {
+    results.add(template);
+
+    const innerTemplates = dom5.queryAll(
+        template.content,
+        matchInnerDataBindingTemplate,
+        [],
+        dom5.childNodesIncludeTemplate) as Template[];
+    for (const innerTemplate of innerTemplates) {
+      results.add(innerTemplate);
+    }
+  }
+
+  return results;
+}
+
+
+/**
+ * Databinding into an attribute/property.
+ */
+export class ScannedAttributeExpression implements ScannedFeature {
+  /** The element whose attribute is being assigned to.*/
+  astNode: parse5.ASTNode;
+  /** The HTML element attribute that's being assigned to.*/
+  attribute: parse5.ASTAttribute;
+  sourceRange: SourceRange;
+  warnings: Warning[] = [];
+  directionality: 'down-only'|'bidirectional';
+  expressionText: string;
+  expressionAst: estree.Program;
+
+  /**
+   * If this is a bidirectional data binding, and an event name was specified
```

Comment:
Doing it this way, you get a javascript parse error if you use `::` on anything but a bidirectional match of an entire attribute. No need for the additional lint pass.